File: src/lib/api/channel.ts
import { invoke } from '@tauri-apps/api/core';
import type { Channel } from '$lib/types';

/**
 * Get all channels for a specific playlist.
 * @param playlistId - The UUID of the playlist
 * @returns Array of Channel objects
 */
export async function getChannelsByPlaylist(playlistId: string): Promise<Channel[]> {
  return await invoke<Channel[]>('get_channels_by_playlist', { playlistId });
}

/**
 * Search channels by name across playlists.
 * @param query - The search query string
 * @param playlistId - Optional playlist ID to limit search to
 * @returns Array of matching Channel objects
 */
export async function searchChannels(
  query: string,
  playlistId?: string
): Promise<Channel[]> {
  return await invoke<Channel[]>('search_channels', { query, playlistId });
}

/**
 * Get a single channel by its ID.
 * @param id - The UUID of the channel
 * @returns The Channel object
 */
export async function getChannelById(id: string): Promise<Channel> {
  return await invoke<Channel>('get_channel_by_id', { id });
}

/** Whether an M3U/Xtream-sourced channel supports catch-up (archive replay) playback at all. */
export async function m3uIsCatchupSupported(channelId: string): Promise<boolean> {
  return await invoke<boolean>('m3u_is_catchup_supported', { channelId });
}

/**
 * Builds a playable catch-up URL for a channel at an EPG program's start.
 * @param programStart - ISO or XMLTV-formatted date string for the program's start
 * @param programStartTimestamp - Unix seconds, when known - takes precedence over `programStart`
 * @returns The playable URL, or null if this channel/program doesn't support catch-up
 */
export async function m3uResolveCatchupUrl(
  channelId: string,
  programStart: string,
  programStartTimestamp?: number
): Promise<string | null> {
  return await invoke<string | null>('m3u_resolve_catchup_url', {
    channelId,
    programStart,
    programStartTimestamp,
  });
}

File: src-tauri/src/commands/channel.rs
use crate::db;
use crate::error::{CommandError, CommandResult};
use crate::state::AppState;
use crate::types::Channel;
use tauri::State;

#[tauri::command]
pub async fn get_channels_by_playlist(
    state: State<'_, AppState>,
    playlist_id: String,
) -> CommandResult<Vec<Channel>> {
    db::with_conn(&state.db, move |conn| Ok(db::channels::list_by_playlist(conn, &playlist_id)?)).await
}

#[tauri::command]
pub async fn search_channels(
    state: State<'_, AppState>,
    query: String,
    playlist_id: Option<String>,
) -> CommandResult<Vec<Channel>> {
    db::with_conn(&state.db, move |conn| {
        Ok(db::channels::search(conn, &query, playlist_id.as_deref())?)
    })
    .await
}

#[tauri::command]
pub async fn get_channel_by_id(state: State<'_, AppState>, id: String) -> CommandResult<Channel> {
    let lookup_id = id.clone();
    db::with_conn(&state.db, move |conn| Ok(db::channels::get_by_id(conn, &lookup_id)?))
        .await?
        .ok_or_else(|| CommandError::NotFound(format!("channel {id} not found")))
}

/// Whether a channel supports catch-up (archive replay) — a cheap gate for
/// the "watch from start" affordance, callable without a specific EPG
/// program in hand.
#[tauri::command]
pub async fn m3u_is_catchup_supported(state: State<'_, AppState>, channel_id: String) -> CommandResult<bool> {
    let lookup_id = channel_id.clone();
    let channel = db::with_conn(&state.db, move |conn| Ok(db::channels::get_by_id(conn, &lookup_id)?))
        .await?
        .ok_or_else(|| CommandError::NotFound(format!("channel {channel_id} not found")))?;
    Ok(crate::catchup::is_m3u_catchup_supported(&channel))
}

/// Builds a playable catch-up URL for `channel_id` at an EPG program's start.
/// `program_start_timestamp` (Unix seconds) takes precedence over parsing
/// `program_start` when both are given. `None` if catch-up isn't supported
/// or the URL can't be built.
#[tauri::command]
pub async fn m3u_resolve_catchup_url(
    state: State<'_, AppState>,
    channel_id: String,
    program_start: String,
    program_start_timestamp: Option<i64>,
) -> CommandResult<Option<String>> {
    let lookup_id = channel_id.clone();
    let channel = db::with_conn(&state.db, move |conn| Ok(db::channels::get_by_id(conn, &lookup_id)?))
        .await?
        .ok_or_else(|| CommandError::NotFound(format!("channel {channel_id} not found")))?;
    Ok(crate::catchup::resolve_m3u_catchup_url(&channel, &program_start, program_start_timestamp, None))
}
